src/irc/PrivacyProtection.ts

```diff
@@ -0,0 +1,161 @@
+                isMissingUsers = true;
+            }
+        }
+        if (!isMissingUsers) {
+            return true;
+        }
+        // For the missing users, attempt to join them to the channel. Any that fail to join should be kicked.
+        this.ircBridge.syncMembersInRoomToIrc(req, roomId, new IrcRoom(server, channel), true);
+        return false;
+    }
+
+    /**
+     * Send a `org.matrix.appservice-irc.connection` state event into the room when a channel
+     * is blocked or unblocked. Subsequent calls with the same state will no-op.
+     * @param req The IRC request
+     * @param roomId The Matrix room
+     * @param channel The IRC room
+     * @param blocked Is the channel blocked
+     * @returns A promise, but it will always resolve.
+     */
+    private async setBlockedStateInRoom(req: BridgeRequest, roomId: string, ircRoom: IrcRoom, blocked: boolean) {
+        const key = roomId + ircRoom.getId();
+        if (this.roomBlockedSet.has(key) === blocked) {
+            return;
+        }
+        if (blocked) {
+            this.roomBlockedSet.add(key);
+            req.log.warn(`${roomId} ${ircRoom.getId()} is now blocking IRC messages`);
+        }
+        else {
+            this.roomBlockedSet.delete(key);
+            req.log.warn(`${roomId} ${ircRoom.getId()} has now unblocked IRC messages`);
+        }
+        try {
+            const intent = this.ircBridge.getAppServiceBridge().getIntent();
+            // This is set *approximately* for when the room is unblocked, as we don't do when a new user joins.
+            await intent.sendStateEvent(roomId, "org.matrix.appservice-irc.connection", ircRoom.getId(), {
+                blocked,
+            });
+        }
+        catch (ex) {
+            req.log.warn(`Could not set org.matrix.appservice-irc.connection in room`, ex);
+        }
+    }
+
+    /**
+     * Get rooms which are safe to bridge IRC messages to.
+     * @param req The bridge request
+     * @param server The IRC server
+     * @param channel The IRC channel
+     * @returns An array of Matrix rooms
+     */
+    async getSafeRooms(req: BridgeRequest, server: IrcServer, channel: string): Promise<MatrixRoom[]> {
+        const allRooms = await this.ircBridge.getStore().getMatrixRoomsForChannel(server, channel);
+        return (await Promise.all((allRooms).map(async (room) => {
+            const required = await this.shouldRequireMatrixUserJoined(server, channel, room.roomId);
+            req.log.debug(`${room.roomId} ${required ? "requires" : "does not require"} Matrix users to be joined`);
+            if (!required) {
+                return room;
+            }
+            const allowed = await this.areAllMatrixUsersJoined(req, server, channel, room.roomId);
+            // Do so asynchronously, as we don't want to block message handling on this.
+            this.setBlockedStateInRoom(req, room.roomId, new IrcRoom(server, channel), !allowed);
```

Comment:
While the function says it's always going to resolve.. there could be a coding error resulting in an unhandled promise.

To log potential errors, implying one can still log to `req.log` after a response has been sent:
```suggestion
            this.setBlockedStateInRoom(req, room.roomId, new IrcRoom(server, channel), !allowed).catch(req.log.error);
```

Reply:
Yep, seems fine to me.